source/js/plugins/_filters.js

```diff
@@ -1,17 +1,83 @@
 
-let _filterQueue = {};
+// stores an array of objects, each object stores the function, image data, and adjust to apply
+let _filterQueue = [];
+// stores whether the invert filter was used (for if it should be reapplied)
+let inverted = false;
 
-export function addFilterToQueue (filter)
+export function resetFilters ()
 {
-    let fname = filter.name;
+    _filterQueue = [];
+    inverted = false;
+}
 
-    if (_filterQueue.hasOwnProperty(fname))
+// add a filter to the array. if it is new, apply the filter's function to the image data of the previous filter's 
+// returned image data (or the default image data if it's the first filter), and return this new image data
+export function addFilterToQueue (filter, data, adjust)
+{
+    // index of the filter in the queue, -1 if not found
+    let index = _filterQueue.findIndex(f => f.filter.name === filter.name);
+    if (index !== -1)
     {
-        // update existing filter
+        // adjust a filter already in the queue
+        let filtObj = _filterQueue[index];
+        filtObj.adjust = adjust;
+        
+        // sharpness filter uses convolve not apply
+        if (filtObj.filter.name === 'convolve') 
+            filtObj.postData = filtObj.filter(filtObj.prevData, filtObj.adjust);
+        else if (filtObj.filter.name === '_invert') // invert filter should toggle, so use post-alteration image data
+        {
+            filtObj.postData = _apply(filtObj.postData, filtObj.filter, filtObj.adjust);
+            inverted = !inverted;
+        }
+        else 
+            filtObj.postData = _apply(filtObj.prevData, filtObj.filter, filtObj.adjust);
+
+        // reapply all filters that come after in the queue
+        for (let i = index + 1, len = _filterQueue.length; i < len; i++)
+        {
+            let otherFiltObj = _filterQueue[i];
+
+            if (otherFiltObj.filter.name === '_invert' && !inverted) // don't reinvert the image
+                continue;
+
+            otherFiltObj.prevData = _filterQueue[i - 1].postData; // starts at filt
+
+            // sharpness filter uses convolve not apply
```

Comment:
why?

Reply:
The sharpness filter doesn't use the `_apply` private function, whereas all the other ones do.